File: resources/recipes/api/input-validation/validate_user.py
"""Input validation — runnable Pydantic v2 example.

Schema-first validation: declare the shape once, get structured
field-level errors for free.

Run:  python validate_user.py
"""
from pydantic import BaseModel, EmailStr, Field, ValidationError, field_validator


class UserCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    email: EmailStr
    age: int = Field(ge=0, le=150)
    bio: str | None = Field(default=None, max_length=500)

    @field_validator("name")
    @classmethod
    def name_must_not_be_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("name cannot be blank")
        return v.strip()


def format_errors(exc: ValidationError) -> list[dict]:
    """Turn ValidationError into field-level error objects for a 400 response."""
    return [
        {"field": ".".join(map(str, e["loc"])), "message": e["msg"], "type": e["type"]}
        for e in exc.errors()
    ]


if __name__ == "__main__":
    # Valid input
    user = UserCreate(name="Ada Lovelace", email="ada@example.com", age=36)
    print("valid:", user.model_dump())

    # Invalid input — all violations at once, not just the first
    try:
        UserCreate(name="  ", email="not-an-email", age=200)
    except ValidationError as e:
        print("invalid:", format_errors(e))

File: resources/recipes/api/input-validation/validate_user.js
/**
 * Input validation — runnable Zod example.
 *
 * Schema-first validation: declare the shape once, use safeParse to
 * collect every violation as field-level errors.
 *
 * Run:  node validate_user.js   (requires: npm install)
 */

const { z } = require('zod');

const UserCreate = z.object({
  name: z.string().trim().min(1, 'name cannot be blank').max(100),
  email: z.string().email(),
  age: z.number().int().min(0).max(150),
  bio: z.string().max(500).optional(),
});

/** Turn a ZodError into field-level error objects for a 400 response. */
function formatErrors(error) {
  return error.issues.map((issue) => ({
    field: issue.path.join('.'),
    message: issue.message,
    code: issue.code,
  }));
}

/** Parse and return { data } or { errors } — never throws. */
function validateUser(payload) {
  const result = UserCreate.safeParse(payload);
  return result.success ? { data: result.data } : { errors: formatErrors(result.error) };
}

module.exports = { UserCreate, validateUser, formatErrors };

if (require.main === module) {
  console.log('valid:', validateUser({ name: 'Ada Lovelace', email: 'ada@example.com', age: 36 }));
  console.log('invalid:', validateUser({ name: '  ', email: 'not-an-email', age: 200 }));
}
